src/snowflake/cli/_plugins/snowpark/commands.py

```diff
@@ -132,133 +140,155 @@ def deploy(
-        project_definition=pd,
+    snowpark_entities = get_snowpark_entities(pd)
+    project_name = (
+        pd.defaults.project_name if pd.defaults and pd.defaults.project_name else ""
+    )
+    project_paths = SnowparkProjectPaths(
+        project_root=cli_context.project_root, project_name=project_name
     )
 
-    procedures: Dict[str, ProcedureEntityModel] = pd.get_entities_by_type("procedure")
-    functions: Dict[str, FunctionEntityModel] = pd.get_entities_by_type("function")
+    with cli_console.phase("Performing initial validation"):
+        if not snowpark_entities:
+            raise ClickException(
+                "No procedures or functions were specified in the project definition."
+            )
+        validate_all_artifacts_exists(
+            project_paths=project_paths, snowpark_entities=snowpark_entities
+        )
 
-    if not procedures and not functions:
-        raise ClickException(
-            "No procedures or functions were specified in the project definition."
+    # Validate current state
+    with cli_console.phase("Checking remote state"):
+        om = ObjectManager()
+        _check_if_all_defined_integrations_exists(om, snowpark_entities)
+        existing_objects = check_for_existing_objects(om, replace, snowpark_entities)
+
+    with cli_console.phase("Preparing required stages and artifacts"):
+        entities_to_imports_map, stages_to_artifact_map = build_artifacts_mappings(
+            project_paths=project_paths,
+            snowpark_entities=snowpark_entities,
         )
 
-    if not paths.artifact_file.exists():
-        raise ClickException(
-            "Artifact required for deploying the project does not exist in this directory. "
-            "Please use build command to create it."
+        create_stages_and_upload_artifacts(stages_to_artifact_map)
+
+    # Create snowpark entities
+    with cli_console.phase("Creating Snowpark entities"):
+        snowflake_dependencies = _read_snowflake_requirements_file(
+            project_paths.snowflake_requirements
         )
+        deploy_status = []
+        for entity in snowpark_entities.values():
+            cli_console.step(f"Creating {entity.type} {entity.fqn}")
+            operation_result = _deploy_single_object(
+                entity=entity,
+                existing_objects=existing_objects,
+                snowflake_dependencies=snowflake_dependencies,
+                entities_to_artifact_map=entities_to_imports_map,
+            )
+            deploy_status.append(operation_result)
 
-    pm = ProcedureManager()
-    fm = FunctionManager()
-    om = ObjectManager()
+    return CollectionResult(deploy_status)
 
-    _check_if_all_defined_integrations_exists(om, functions, procedures)
 
-    existing_functions = _find_existing_objects(ObjectType.FUNCTION, functions, om)
-    existing_procedures = _find_existing_objects(ObjectType.PROCEDURE, procedures, om)
+def validate_all_artifacts_exists(
```

Comment:
Do we want to check for at least one of `dependencies.zip`/`requirements.snowflake.txt` in case `requirements.txt` exist?

Reply:
We don't do so today, I will consider this as a feature request that I can add in followup PR